src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+  }
+
+  void analyzePublicTypes(Module& wasm) {
+    // We cannot change supertypes for anything public.
+    for (auto type : ModuleUtils::getPublicHeapTypes(wasm)) {
+      if (auto super = type.getSuperType()) {
+        noteSubtype(type, *super);
+      }
+    }
+  }
+
+  void analyzeTransitiveDependencies() {
+    // While we have found new subtypings and have not reached a fixed point...
+    while (!worklist.empty()) {
+      // Subtype relationships that we are keeping might depend on other subtype
+      // relationships that we are not yet planning to keep. Transitively find
+      // all the relationships we need to keep all our type definitions valid.
+      while (!worklist.empty()) {
+        auto it = worklist.begin();
+        auto type = *it;
+        worklist.erase(it);
+        auto super = supertypes.at(type);
+        if (super.isBasic()) {
+          continue;
+        }
+
+        if (type.isStruct()) {
+          const auto& fields = type.getStruct().fields;
+          const auto& superFields = super.getStruct().fields;
+          for (size_t i = 0, size = superFields.size(); i < size; ++i) {
+            if (fields[i].mutable_ == Immutable) {
+              noteSubtype(fields[i].type, superFields[i].type);
+            }
+          }
+        } else if (type.isArray()) {
+          auto elem = type.getArray().element;
+          if (elem.mutable_ == Immutable) {
+            noteSubtype(elem.type, super.getArray().element.type);
+          }
+        } else {
+          assert(type.isSignature());
+          auto sig = type.getSignature();
+          auto superSig = super.getSignature();
+          noteSubtype(superSig.params, sig.params);
+          noteSubtype(sig.results, superSig.results);
+        }
+      }
+
+      // Analyze all casts at once.
+      // TODO: This is expensive. Analyze casts incrementally after we
+      // initially analyze them.
+      analyzeCasts();
+    }
+  }
+
+  void analyzeCasts() {
+    // For each cast (src, dest) pair, any type that remains a subtype of src
+    // (meaning its values can inhabit locations typed src) and that was
+    // originally a subtype of dest (meaning its values would have passed the
+    // cast) should remain a subtype of dest so that its values continue to pass
+    // the cast.
+    //
+    // For every type, walk up its new supertype chain to find cast sources and
+    // compare against their associated cast destinations.
+    for (auto it = supertypes.begin(); it != supertypes.end(); ++it) {
+      auto type = it->first;
+      for (auto superIt = it; superIt != supertypes.end();
+           superIt = supertypes.find(superIt->second)) {
+        auto super = superIt->second;
+        auto destsIt = castTypes.find(super);
+        if (destsIt == castTypes.end()) {
+          continue;
+        }
+        for (auto dest : destsIt->second) {
+          if (HeapType::isSubType(type, dest)) {
+            noteSubtype(type, dest);
```

Comment:
`dest` is used in the code, but the comment has `dest, src`, so it would be good to use `src` in the code as well to match the comment.

Reply:
Sure, but it's `super` that will be renamed to `dest`, not `type`. If that's surprising, then that further motivates aligning the names.